tests/test_warnings.cpp

```diff
@@ -0,0 +1,79 @@
+/*
+    tests/test_warnings.cpp -- usage of warnings::warn() and warnings categories.
+
+    Copyright (c) 2024 Jan Iwaszkiewicz <jiwaszkiewicz6@gmail.com>
+
+    All rights reserved. Use of this source code is governed by a
+    BSD-style license that can be found in the LICENSE file.
+*/
+
+#include <pybind11/warnings.h>
+
+#include "pybind11_tests.h"
+
+#include <utility>
+
+namespace warning_helpers {
+void warn_function(py::module &m, const char *name, py::handle category, const char *message) {
+    m.def(name, [category, message]() { py::warnings::warn(message, category); });
+}
+} // namespace warning_helpers
+
+class CustomWarning {};
+
+TEST_SUBMODULE(warnings_, m) {
+
+    // Test warning mechanism base
+    m.def("raise_and_return", []() {
+        std::string message = "Warning was raised!";
+        py::warnings::warn(message.c_str(), PyExc_Warning);
+        return 21;
+    });
+
+    m.def("raise_default", []() { py::warnings::warn("RuntimeWarning is raised!"); });
+
+    m.def("raise_from_cpython",
+          []() { py::warnings::warn("UnicodeWarning is raised!", PyExc_UnicodeWarning); });
+
+    m.def("raise_and_fail",
+          []() { py::warnings::warn("RuntimeError should be raised!", PyExc_Exception); });
+
+    // Test custom warnings
+    PYBIND11_CONSTINIT static py::gil_safe_call_once_and_store<py::object> ex_storage;
+    ex_storage.call_once_and_store_result([&]() {
+        return py::warnings::new_warning_type(m, "CustomWarning", PyExc_DeprecationWarning);
+    });
+
+    m.def("raise_custom", []() {
+        py::warnings::warn("CustomWarning was raised!", ex_storage.get_stored());
+        return 37;
+    });
+
+    m.def("register_duplicate_warning",
+          [m]() { py::warnings::new_warning_type(m, "CustomWarning", PyExc_RuntimeWarning); });
+
+    // Bind warning categories
+    warning_helpers::warn_function(m, "raise_base_warning", PyExc_Warning, "This is Warning!");
+    warning_helpers::warn_function(
+        m, "raise_bytes_warning", PyExc_BytesWarning, "This is BytesWarning!");
+    warning_helpers::warn_function(
+        m, "raise_deprecation_warning", PyExc_DeprecationWarning, "This is DeprecationWarning!");
```

Comment:
I'd delete all warning below: mostly because it doesn't exercise anything specific in your implementation, but also because it will look old after a few years, because we're likely to not keep up with new warning types, and the wall of repetitive code is distracting from potentially more important things.

I think it's best to only keep `PyExc_Warning` (the base class) and one randomly picked derived class.

Reply:
I have removed them all as other functions like `warn_and_return_value` or `warn_with_different_category` use these categories. This also result in obsolecence of `namespace pybind11_tests/warnings` as no helpers are needed now (ref https://github.com/pybind/pybind11/pull/5291#discussion_r1719965503).